### PR TITLE
PERF: tighten _should_compare for MultiIndex

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -5289,6 +5289,16 @@ def _get_indexer_non_comparable(
         """
         if method is not None:
             other = unpack_nested_dtype(target)
+            if self._is_multi ^ other._is_multi:
+                kind = other.dtype.type if self._is_multi else self.dtype.type
```
Comment (jreback) [inline pandas/core/indexes/base.py:5293]: do tests hit this? e.g. as you didn't change anything

**Comment by jreback** [comment]
is there a specific benchmark that this improves?

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py:5349]: One of the affected cases was not tested; just added a test for that case.

**Comment by jbrockmendel** [comment]
> is there a specific benchmark that this improves?

No.  The motivation is yak-shaving that traces back to getting rid of _convert_list_indexer

**Comment by jbrockmendel** [comment]
@jreback gentle ping (a whole mess of MultiIndex PRs yak-shaving inconsistencies)

**Comment by bashtage** [comment]
@jbrockmendel  This PR seems to have introduced a bug. I verified this with the code below and a bisect.

```
import pandas as pd
mi = pd.MultiIndex.from_product([["a","b","c"],[1,2,3],["z","y","x"]])
df = pd.DataFrame(index=mi,dtype=float)
mi2 = pd.MultiIndex.from_product([["a","b","c"],[1,2,3]])
s = pd.Series(index=mi2, dtype=float)
s.iloc[:]=3.14
df["new"] = s
print(df)
```

now returns 
```
        new
a 1 z  NaN
    y  NaN
    x  NaN
  2 z  NaN
    y  NaN
    x  NaN
  3 z  NaN
    y  NaN
    x  NaN
```

Before this patch is did a broadcast assignment to the remaining MultiIndex levels, i.e..


```
        new
a 1 z  3.14
    y  3.14
    x  3.14
  2 z  3.14
    y  3.14
    x  3.14
  3 z  3.14
    y  3.14
    x  3.14
```

**Comment by bashtage** [comment]
xref #40186

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -5474,6 +5484,14 @@ def _should_compare(self, other: Index) -> bool:
 
         other = unpack_nested_dtype(other)
         dtype = other.dtype
+        if other._is_multi:
+            if not self._is_multi:
+                # other contains only tuples so unless we are object-dtype,
+                #  there can never be any matches
+                return self._is_comparable_dtype(dtype)
+            return self.nlevels == other.nlevels
```
Comment (bashtage) [inline pandas/core/indexes/base.py:5492]: This is the change that is breaking MultiIndex broadcasting.  If one has 3 levels and the other has 2, then this is False.  Previously these were comparable and so would be compared and expanded.

Comment (jbrockmendel) [reply pandas/core/indexes/base.py:5492]: thanks.  do you know what the calling method is in the problematic case?

Comment (bashtage) [reply pandas/core/indexes/base.py:5492]: Walking back, previous is

https://github.com/pandas-dev/pandas/blob/ddd90b0f171fac3c2e39bc2ff76df97d57f36a18/pandas/core/indexes/base.py#L3481

then 

https://github.com/pandas-dev/pandas/blob/ddd90b0f171fac3c2e39bc2ff76df97d57f36a18/pandas/core/indexes/base.py#L3887

Here `self` is the Series with 2 levels and other is the DataFrame with 3.

Comment (bashtage) [reply pandas/core/indexes/base.py:5492]: The big change is driven by the return difference of `self._should_compare(target)`. Before this patch it returned `True`, so the `if not ...` block was skipped.  It now returns `False`, and so it incorrectly shortcuts and fills with an NA value.

Comment (jbrockmendel) [reply pandas/core/indexes/base.py:5492]: OK, i think ive got a handle on whats going on here.  The long-term fix will be in MultiIndex.get_indexer, but for now this should just be reverted.

**Pull request merged**